### PR TITLE
Create dry-run mode and respect dry run for git push calls

### PR DESCRIPTION
**Pull request by josephperrott**
See individual commits.

### CODE REVIEW

**Review by devversion — COMMENTED** [review]
I was thinking a lot about adding a dry run mode to the release script but never came to a good solution. Are we sure this is the solution we'd want to go with dry-run for commands in dev-infra? My concern is that we are adding magic logic within `GitClient` for dry-run, while each dry-run could/should be ("through-through") differently per ng-dev command. 

I'd actually expect specific dry-run logic per ng-dev command (e.g. the release tool).

<sup>One could argue that `git push` should never execute if dry-run is enabled; but I'm rather disliking the fact that this happens magically within the `GitClient`</sup>.

**Comment by josephperrott** [comment]
Yeah, I worry about that a bit as well.  And I couldn't really decide where to fall on whether dry run should be done per command entirely, or have some underlying structure pieces.

My thought here was that rather than having to pass this state through a bunch of places where it is unused, we could have the state be able to be checked in locations where it actually matters.  So the `GitClient` will ensure that even if the dry-run logic for a specific command still allows `git push` to be called, it will be a noop.

I am not entirely sure though.

**Comment by josephperrott** [comment]
@devversion PTAL.  Updated to throw an error rather than noop when `push` is used in dryRun mode.

File: dev-infra/utils/dry-run.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {Argv} from 'yargs';
+
+/**
+ * Add a --dry-run flag to the available options for the yargs argv object.  When present, sets an
```
Comment (devversion) [inline dev-infra/utils/dry-run.ts:12]: Nit: Extra space

**Review by devversion — APPROVED** [review]
LGTM!

**Pull request closed without merging**